src/controllers/hid/hidioglobaloutputreportfifo.cpp

```diff
@@ -0,0 +1,93 @@
+#include "controllers/hid/hidioglobaloutputreportfifo.h"
+
+#include <hidapi.h>
+
+#include "controllers/defs_controllers.h"
+#include "controllers/hid/legacyhidcontrollermappingfilehandler.h"
+#include "util/compatibility/qbytearray.h"
+#include "util/compatibility/qmutex.h"
+#include "util/string.h"
+#include "util/time.h"
+#include "util/trace.h"
+
+namespace {
+constexpr int kMaxHidErrorMessageSize = 512;
+constexpr int kSizeOfFifoInReports = 32;
+} // namespace
+
+HidIoGlobalOutputReportFifo::HidIoGlobalOutputReportFifo()
+        : m_fifoQueue(kSizeOfFifoInReports) {
+}
+
+void HidIoGlobalOutputReportFifo::addReportDatasetToFifo(const quint8 reportId,
+        const QByteArray& reportData,
+        const mixxx::hid::DeviceInfo& deviceInfo,
+        const RuntimeLoggingCategory& logOutput) {
+    // First byte must always be the ReportID-Byte
+    QByteArray report(reportData);
+    report.prepend(reportId); // In Qt6 this is a very fast operation without reallocation
+
+    // Swap report to lockless FIFO queue
+    bool success = m_fifoQueue.try_emplace(std::move(report));
+
+    // Handle the case, that the FIFO queue is full - which is an error case
+    if (!success) {
+        // If the FIFO is full, we skip the report dataset even
+        // in non-skipping mode, to keep the controller mapping thread
+        // responsive for InputReports from the controller.
+        // Alternative would be to block processing of the controller
+        // mapping thread, until the FIFO has space again.
+        qCWarning(logOutput)
+                << "FIFO overflow: Unable to add OutputReport " << reportId
+                << "to the global cache for non-skipping sending of OututReports for"
+                << deviceInfo.formatName();
+    }
+}
+
+bool HidIoGlobalOutputReportFifo::sendNextReportDataset(QMutex* pHidDeviceAndPollMutex,
+        hid_device* pHidDevice,
+        const mixxx::hid::DeviceInfo& deviceInfo,
+        const RuntimeLoggingCategory& logOutput) {
+    auto startOfHidWrite = mixxx::Time::elapsed();
+
+    auto pFront = m_fifoQueue.front();
+
+    if (pFront == nullptr) {
+        // No data in FIFO to be send
+        // Return with false, to signal the caller, that no time consuming IO
+        // operation was ncessary
+        return false;
+    }
+
+    QByteArray dataToSend(std::move(*pFront));
```

Comment:
dataToSend has the report ID prepended. This is not clear initially. 
Can we improve the situation by always putting the report ID at the beginning? 
This will also solve the reallocation of the buffer when prepending the report ID before putting it into the buffer. 

The same benefit goes for 
void HidController::send(QByteArray data, unsigned int reportID) 
The we just not store the modified buffer.

Reply:
Our Javascript API has seperated arguments for ReportID and report data. This is not only needed for backwards compatibility, but is also the cleanest way to define such an API.
I will improve naming and add a comment, to make clear that this variable contains ReportID + data.